bleak/backends/scanner.py

```diff
@@ -61,6 +61,11 @@ def __repr__(self) -> str:
     Optional[Awaitable[None]],
 ]
 
+AdvertisementDataFilter = Callable[
+    [BLEDevice, Optional[AdvertisementData]],
```

Comment:
I don't thing `AdvertisementData` should be `Optional`. Am I missing something?

Reply:
The filter function is called in two places: once from the discovery callback but then later when filtering the results for returning to the caller. The second time the advertisement data isn't available. I see various workable solutions, please advise:
1. The current situation where (the the filter function is called once with 2 args once with 1 arg)
2. Getting rid of the advertisement data arg to the filter function (but that means you can no longer filter on interesting things in the advertisement data, which may be useful in some cases)
3. Getting rid of the filtering of the scan results and simply returning the first result (possibly raising an exception if more than one result was somehow available, I'm not 100% sure this can't happen with all the async stuff going on)
4. Change the signature/semantics of the filter function, stating that it returns either `None` or the `BLEDevice` that should be returned from `find_device_by_filter`, but I don't know whether it is possible to do more in the callback routine than simply setting a flag to make the discovery stop.

Reply:
My choice would be 3. What I usually do is create an `asyncio.Queue()` ([example](https://github.com/hbldh/bleak/blob/develop/examples/uart_service.py)). This allows the callback to queue more than one match in case it receives more before we can stop the scanner but we only use the first item from the queue since it was the first match.

Reply:
Ok, for this use case I'll just do the simple thing and return the first match (after the scan has been stopped), or `None` if there's no matches.

But your `asyncio.Queue()` example makes me wonder: couldn't the whole discovery process be converted to using these? That way the whole filtering stuff would become a lot easier, and various other things would be possible with discovery too, as the results would become available as they were received...

Reply:
Implemented (3).